File: lua/problem_38.lua
local function nQueens(n)
    local cols  = {}
    local diag1 = {}
    local diag2 = {}
    local count = 0
    local function place(row)
        if row > n then count = count + 1; return end
        for col = 1, n do
            local d1 = row - col
            local d2 = row + col
            if not cols[col] and not diag1[d1] and not diag2[d2] then
                cols[col] = true; diag1[d1] = true; diag2[d2] = true
                place(row + 1)
                cols[col] = nil; diag1[d1] = nil; diag2[d2] = nil
            end
        end
    end
    place(1)
    return count
end

local function problem38()
    print("Problem 38:")
    assert(nQueens(1) == 1,  "assertion failed")
    assert(nQueens(4) == 2,  "assertion failed")
    assert(nQueens(8) == 92, "assertion failed")
    print("All Tests Passed")
end

return { problem38 = problem38 }

File: go/problem_38.go
package main

import "fmt"

// nQueens returns the number of valid N-queens arrangements.
func nQueens(n int) int {
	cols := make([]bool, n)
	diag1 := make([]bool, 2*n)
	diag2 := make([]bool, 2*n)
	count := 0
	var place func(row int)
	place = func(row int) {
		if row == n { count++; return }
		for col := 0; col < n; col++ {
			if cols[col] || diag1[row-col+n] || diag2[row+col] { continue }
			cols[col] = true; diag1[row-col+n] = true; diag2[row+col] = true
			place(row + 1)
			cols[col] = false; diag1[row-col+n] = false; diag2[row+col] = false
		}
	}
	place(0)
	return count
}

func problem38() {
	fmt.Println("Problem 38:")
	if nQueens(1) != 1  { panic("assertion failed") }
	if nQueens(4) != 2  { panic("assertion failed") }
	if nQueens(8) != 92 { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}
